### ISSUE TITLE
Block Populator 1.13 crash

### ISSUE THREAD

**Issue by smilesdc** [PaperMC/Paper#1493]
### Link/paste of stack trace
[latest.log](https://github.com/PaperMC/Paper/files/2421495/latest.log)
### Plugin list:
GiantCaves
### Actions to reproduce (if known):
I updated GiantCaves plugin from 1.10 version to 1.13.
https://github.com/rmichela/GiantCaves
This is a populator class:
```
public class GiantCavePopulator extends BlockPopulator {

    private final BlockFace[] faces = { BlockFace.UP, BlockFace.NORTH, BlockFace.SOUTH, BlockFace.EAST, BlockFace.WEST };

    private final Config config;
    private final Plugin plugin;

    private final Material material;
    private final BlockToucher toucher;

    public GiantCavePopulator(Plugin plugin, Config config) {
        this.config = config;
        this.plugin = plugin;
        material = Material.AIR;
        toucher = new BlockToucher(plugin);
        plugin.getServer().getPluginManager().registerEvents(new GCWaterHandler(config), plugin);
    }

    @Override
    public void populate(final World world, final Random random, final Chunk source) {
        GCRandom gcRandom = new GCRandom(source, config);

        for (int x = 0; x < 16; x++) {
            for (int z = 0; z < 16; z++) {
                for (int y = config.caveBandMax; y >= config.caveBandMin; y--) {
                    if (gcRandom.isInGiantCave(x, y, z)) {
                        Block block = source.getBlock(x, y, z);
                        if (isHoldingBackOcean(block)) {
                            // Support the ocean with stone to keep the bottom from falling out
                            if (block.getType().hasGravity()) {
                                block.setType(Material.STONE);
                            }
                        } else {
                            block.setType(material);

                            // Mark adjacent blocks for update, iff they are not in the cave
                            for (BlockFace direction : faces) {
                                Block b = block.getRelative(direction);
                                if (isWater(b) || isLava(b)) {
                                    if (!gcRandom.isInGiantCave(b.getX(), b.getY(), b.getZ())) {
                                        toucher.touch(b);
                                    }
                                }
                            }

                            if (config.debugMode) {
                                block = source.getBlock(x, 192, z);
                                block.setType(Material.GLASS);
                            }
                        }
                    }
                }
            }
        }
    }

    private boolean isHoldingBackOcean(Block block) {
        return isSurfaceWater(block) || isNextToSurfaceWater(block);
    }

    private boolean isNextToSurfaceWater(Block block) {
        for (BlockFace face : faces) {
            Block adjacent = block.getRelative(face);
            if (isSurfaceWater(adjacent)) {
                return true;
            }
        }
        return false;
    }

    private boolean isSurfaceWater(Block block) {
        // Walk the column of blocks above block looking sea level
        while (isWater(block)) {
            if (block.getY() >= block.getWorld().getSeaLevel() - 1) {
                return true;
            } else {
                block = block.getRelative(BlockFace.UP);
            }
        }
        return false;
    }

    private boolean isWater(Block block) {
        Material material = block.getType();
        return material == Material.WATER;
    }

    private boolean isLava(Block block) {
        Material material = block.getType();
        return material == Material.LAVA;
    }
}
```

Server crashes when new chunks generating.

Tested on 1.12.1 version and it's works fine!

While I was debugging plugin I was able to understand that crash occurs when Y exceeds the range from 0 to 15 or idk..

### Paper build number:
This server is running Paper version git-Paper-300 (MC: 1.13.1) (Implementing API version 1.13.1-R0.1-SNAPSHOT)

**Comment by hugmanrique** [comment]
```
DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---
[13:33:38] [Paper Watchdog Thread/ERROR]: The server has not responded for 10 seconds! Creating thread dump
```

**Comment by smilesdc** [comment]
> ```
> DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---
> [13:33:38] [Paper Watchdog Thread/ERROR]: The server has not responded for 10 seconds! Creating thread dump
> ```
Yes I've seen it but I think it's a bug in the core cuz on 1.12.1 version it's working fine.

**Comment by electronicboy** [comment]
Looks like you're crossing chunk boundaries and causing chunk loads, which then triggers your chunk populator which then crosses chunk boundaries and causes a chunk to load

**Comment by smilesdc** [comment]
> Looks like you're crossing chunk boundaries and causing chunk loads, which then triggers your chunk populator which then crosses chunk boundaries and causes a chunk to load

I deleted all water checks and directions check to avoid infinite loop chunks loading and populators run like you said but it's not solve the issue.

On
`block.setType(material);`
Server dies.

I found a solution but I'm not sure if it's correct.
The solution is:
I added 5 ticks delay before changing the world and now it's works.
It looks like there is access block on map or something like that. I don't know.
